### PR TITLE
CU-8669nuweb/As-a-user-when-I-visit-the-system-it-should-automatically-detect-my-location

### PR DESCRIPTION
**Pull request by peter-emad99**
As-a-user-when-I-visit-the-system-it-should-automatically-detect-my-location

### CODE REVIEW

**Comment by peter-emad99** [comment]
Task linked: [CU-8669nuweb As a user, when I visit the system, it should automatically detect my location](https://app.clickup.com/t/8669nuweb)

File: app/Http/Middleware/LocationDetector.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure(\Illuminate\Http\Request): (\Illuminate\Http\Response|\Illuminate\Http\RedirectResponse)  $next
+     * @return \Illuminate\Http\Response|\Illuminate\Http\RedirectResponse
+     */
+    public function handle(Request $request, Closure $next)
+    {
+        $user = $request->user();
+        $userClintIp = \Request::ip();
+        if ($userClintIp == '127.0.0.1')
```
Comment (mahmoud-birdsol) [inline app/Http/Middleware/LocationDetector.php:22]: Please put the braces this does not comply with PSR, also make sure to run `composer pint` before you commit

File: app/Http/Middleware/LocationDetector.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            $userClintIp = '24.48.0.2';
+        if ($user->last_known_ip != $userClintIp || $user->last_known_ip == null) {
+            $userGeolocationData = \App\Services\Geolocator\IpInfo::locate($userClintIp);
+            $user->setAttribute('last_known_ip', $userGeolocationData->ip());
```
Comment (mahmoud-birdsol) [inline app/Http/Middleware/LocationDetector.php:26]: Just use a `$user->update([]);` or `$user->forceFill([])->save();`

File: app/Nova/User.php [hunk 3]
```diff
@@ -194,6 +194,42 @@ public function fields(NovaRequest $request)
                     ->nullable()
                     ->rules('nullable'),
             ]),
+            Text::make('last_known_ip')
```
Comment (mahmoud-birdsol) [inline app/Nova/User.php:197]: In nova use regular field names ```Text::make('Last known ip')```

File: app/Services/Geolocator/IPGeolocationApi.php [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+class IPGeolocationApi implements GeoLocatorApiInterface
+{
+    private  $ip;
+    public function __construct($ip)
+    {
+        $this->ip = $ip;
+    }
+    public  function getDetails()
+    {
+        $ipInfo = Http::get("http://ip-api.com/json/{$this->ip}");
```
Comment (mahmoud-birdsol) [inline app/Services/Geolocator/IPGeolocationApi.php:16]: Move this to the services config file and add it to the .env file with this value as the default in the services config file services.php

File: database/factories/UserFactory.php [hunk 5]
```diff
@@ -58,6 +58,13 @@ public function definition()
             'accepted_cookie_policy_at' => now(),
             'preferred_foot' => $this->faker->randomElement(['left', 'right']),
             'rating' => $this->faker->randomFloat(2, 0, 5),
+            'last_known_ip' => $this->faker->ipv4(),
+            'current_country' => $this->faker->country(),
+            'current_region' => $this->faker->region,
```
Comment (mahmoud-birdsol) [inline database/factories/UserFactory.php:63]: Make this a method not an attribute faker attributes will be deprecated in future versions.

**Review by mahmoud-birdsol — APPROVED** [review]
@ahmy1996404  Some minor comments here please fix then merge the code.

**Pull request merged**